### PR TITLE
chore: dail observed address use another way

### PR DESCRIPTION
**Pull request by driftluo**
There is a modification conflict with #2565

### CODE REVIEW

File: network/src/network.rs [hunk 1]
```diff
@@ -468,31 +469,46 @@ impl NetworkState {
...
                 p2p_control,
                 self.local_peer_id(),
-                addr,
+                addr.clone(),
                 TargetProtocol::Single(SupportProtocols::Identify.protocol_id()),
                 true,
             ) {
-                debug!("try_dial_observed_addrs error {}", err);
+                Err(err) => {
+                    debug!("try_dial_observed_addrs error {}", err);
```
Comment (yangby-cryptape) [inline network/src/network.rs:484]: Just a thinly veiled criticism, the key word `error` in `debug` logs is terrible.
- It makes `error` to be worthless.
  When I enable `log=trace` to debug, then `grep -i error` is meaningless.
  I have to use a complex filter on a huge file, and wait more time.

- Cause panic to users who enable `log=debug`.
  Not all users could distinguish which `error` are insignificant and which are urgent. 
  Yes, I can filter the log records by log level, but I have to use a more complex logs filter, which is terrible with a huge file.

File: network/src/network.rs [hunk 2]
```diff
@@ -591,7 +607,15 @@ impl<T: ExitHandler> ServiceHandle for EventHandler<T> {
...
                 {
                     debug!("dial observed address success: {:?}", address);
-                    public_addrs.insert(addr);
+                    match multiaddr_to_socketaddr(&addr) {
+                        Some(ip) => {
+                            if is_reachable(ip.ip()) {
+                                public_addrs.insert(addr);
+                            }
+                        }
+                        None => (),
```
Comment (yangby-cryptape) [inline network/src/network.rs:616]: Maybe `if ... let ...` is better?

**Review by yangby-cryptape — APPROVED** [review]

**Review by yangby-cryptape — APPROVED** [review]

**Review by yangby-cryptape — APPROVED** [review]

File: network/src/network.rs [hunk 3]
```diff
@@ -468,31 +469,43 @@ impl NetworkState {
     /// this method is intent to check observed addr by dial to self
     pub(crate) fn try_dial_observed_addrs(&self, p2p_control: &ServiceControl) {
         let mut pending_observed_addrs = self.pending_observed_addrs.write();
-        let public_addrs = { self.public_addrs.read().clone() };
-        for addr in pending_observed_addrs
-            .drain()
-            .chain(public_addrs.into_iter())
-        {
+        if pending_observed_addrs.is_empty() {
+            let addrs = self.public_addrs.write();
```
Comment (TheWaWaR) [inline network/src/network.rs:473]: Why dial `self.public_addrs` in a `fn try_dial_observed_addrs()` ?

Comment (driftluo) [reply network/src/network.rs:473]: the observed address is not always valid. For example, the address obtained through the reuse port extension may become invalid as the connection is disconnected. This part of the address information needs to be cleaned up.

File: network/src/network.rs [hunk 4]
```diff
@@ -426,30 +427,43 @@ impl NetworkState {
...
+            let addr = { pending_observed_addrs.iter().next().unwrap().clone() };
+            trace!("try dial observed addr: {:?}", addr);
+            match self.dial_inner(
+                p2p_control,
+                addr.clone(),
+                TargetProtocol::Single(SupportProtocols::Identify.protocol_id()),
+                true,
+            ) {
+                Err(err) => {
+                    trace!("try_dial_observed_addrs fail {} on pending observed", err);
```
Comment (quake) [inline network/src/network.rs:454]: should remove the addr from `pending_observed_addrs` when dial error?

Comment (driftluo) [reply network/src/network.rs:454]: > should remove the addr from `pending_observed_addrs` when dial error?

I think this error is generally that the peerid already exists in the dial list, not tentacle dial queue full, so I think it’s not necessary to delete it here

Comment (driftluo) [reply network/src/network.rs]: The latest commit deals with this problem in a more understandable way

File: network/src/network.rs [hunk 5]
```diff
@@ -426,30 +424,37 @@ impl NetworkState {
...
+        if pending_observed_addrs.is_empty() {
+            let addrs = self.public_addrs.read();
+            if addrs.is_empty() {
+                return;
+            }
+            // random get addr
+            let index = rand::random::<usize>() % addrs.len();
+            let addr = addrs.iter().nth(index).unwrap();
+            if let Err(err) = p2p_control.dial(
+                addr.clone(),
```
Comment (quake) [inline network/src/network.rs:436]:
```suggestion
            if let Some(addr) = addrs.iter().choose(&mut rand::thread_rng()) {
                if let Err(err) = p2p_control.dial(
                    addr.clone(),
                    TargetProtocol::Single(SupportProtocols::Identify.protocol_id()),
                ) {
                    trace!("try_dial_observed_addrs fail {} on public address", err)
                }
            }
```

**Review by quake — DISMISSED** [review]

**Review by quake — APPROVED** [review]

**Comment by nervos-bot-user** [comment]
Can one of the admins verify this patch?

**Review by quake — APPROVED** [review]

**Review by quake — APPROVED** [review]

**Review by doitian — APPROVED** [review]

**Comment by driftluo** [comment]
bors r=quake,doitian

**Pull request merged**